utils/src/linked_list.c

```diff
@@ -0,0 +1,230 @@
+#include "linked_list.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+/**
+ * @brief      Prints data of a linked list.
+ *
+ * @param[in]      list  The list to print
+ */
+void print_linked_list(linked_list_t *list) {
+    // Check for NULL
+    if (list == NULL) {
+        return;
+    }
+    node_t *curr_node = NULL;
+    for (curr_node = list->head; curr_node != NULL;
+         curr_node = curr_node->next) {
+        printf("%s\n", (char *)curr_node->data);
+    }
+}
+
+/**
+ * @brief      Initializes the linked list
+ *
+ * @return     Returns an empty linked list
+ */
+linked_list_t *init_list() {
+    linked_list_t *list = (linked_list_t *)calloc(1, sizeof(linked_list_t));
+
+    // Check NULL
+    if (list == NULL) {
+        return NULL;
+    }
+    return list;
+}
+
+/**
+ * @brief      Returns the data from the tail of the list
+ *
+ * @param[in]      list  The list
+ *
+ * @return     Returns the last item in the list
+ */
+void *get_tail(linked_list_t *list) {
+    // Check for NULL
+    if (list == NULL) {
+        return NULL;
+    }
+
+    // Return tail data if not NULL
+    if (list->tail != NULL) {
+        return list->tail->data;
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief      Insert data into the given linked list
+ *
+ * @param[in]  list  The list for insertion
+ * @param[in]  data  The data to insert
+ *
+ * @return     Returns the node that the data was inserted into else NULL if not
+ * found
+ */
+node_t *insert_data(linked_list_t *list, void *data) {
+    // Check for NULL
+    if (list == NULL) {
+        return NULL;
+    }
+
+    node_t *new_node = (node_t *)calloc(1, sizeof(node_t));
+
+    // Check for NULL
+    if (new_node == NULL) {
+        return NULL;
+    }
+
+    if (list->head == NULL) {
+        // Make this entry the head
+        list->head = new_node;
+        list->head->next = NULL;
+        list->head->data = data;
+        list->tail = list->head;
+    } else {
+        // Insert as new head
+        node_t *temp = list->head;
+        new_node->data = data;
+        new_node->next = temp;
+        list->head = new_node;
+    }
+
+    return new_node;
+}
+
+/**
+ * @brief      Removes the given data from the list; does not free the data
+ *
+ * @param[in]  list  The list for removal
+ * @param[in]  data  The data to remove
+ *
+ * @return     Returns a pointer to the data that was removed, else NULL if not
+ * found
+ */
+void *remove_data(linked_list_t *list, void *data) {
```

Comment:
Only having a function to remove a node based on the data seems kind of limiting.  If the same data is entered into the list twice, there is no way to remove the first entry without removing the second.  Perhaps another function that removes a given node?  This can always be added later anyway.

Reply:
True, we can add a function to remove a given node later

Reply:
https://github.com/bama4/simple-http-server/issues/17